pkg/controller/gardenletdeployer/actuator.go

```diff
@@ -429,64 +432,102 @@ func (a *Actuator) checkSeedSpec(ctx context.Context, spec *gardencorev1beta1.Se
+	)
+
+	// Get backup secret
+	backupSecret, originalErr := kubernetesutils.GetSecretByObjectReference(ctx, a.GardenClient, spec.Backup.CredentialsRef)
+
+	if client.IgnoreNotFound(originalErr) != nil {
+		return originalErr
+	}
+
+	if apierrors.IsNotFound(originalErr) && !allowCopying {
+		return fmt.Errorf("the configured backup secret does not exist, however the feature gate DoNotCopyBackupCredentials is enabled and shoot infrastructure credentials will not be reused: %w", originalErr)
+	}
+
+	if originalErr == nil {
+		checksum = utils.ComputeSecretChecksum(backupSecret.Data)[:8]
+	}
+
+	const (
+		secretStatusLabelKey = "secret.backup.gardener.cloud/status"
+		secretStatusLabelVal = "previously-managed"
+	)
+
+	// Create or update backup secret if it doesn't exist, or is owned by the object, or was previously managed by this controller
+	if allowCopying && (apierrors.IsNotFound(originalErr) || metav1.IsControlledBy(backupSecret, obj) || backupSecret.Labels[secretStatusLabelKey] == secretStatusLabelVal) {
+		gvk, err := apiutil.GVKForObject(obj, a.GardenClient.Scheme())
+		if err != nil {
+			return fmt.Errorf("could not get GroupVersionKind from object %v: %w", obj, err)
+		}
+
+		infrastructureSecret, err := a.GetInfrastructureSecret(ctx)
+		if err != nil {
+			return err
 		}
 
-		// Create or update backup secret if it doesn't exist or is owned by the object
-		if apierrors.IsNotFound(originalErr) || metav1.IsControlledBy(backupSecret, obj) {
-			gvk, err := apiutil.GVKForObject(obj, a.GardenClient.Scheme())
-			if err != nil {
-				return fmt.Errorf("could not get GroupVersionKind from object %v: %w", obj, err)
+		// If there is no infrastructure Secret, e.g. WorkloadIdentity is used instead
+		// we skip the copying as it is not supported.
+		if infrastructureSecret != nil {
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{Namespace: spec.Backup.CredentialsRef.Namespace, Name: spec.Backup.CredentialsRef.Name},
 			}
 
-			infrastructureSecret, err := a.GetInfrastructureSecret(ctx)
-			if err != nil {
+			if _, err := controllerutils.CreateOrGetAndStrategicMergePatch(ctx, a.GardenClient, secret, func() error {
+				secret.OwnerReferences = []metav1.OwnerReference{
+					*metav1.NewControllerRef(obj, gvk),
+				}
+				secret.Type = corev1.SecretTypeOpaque
+				secret.Data = infrastructureSecret.Data
+				delete(secret.Labels, secretStatusLabelKey)
+				return nil
+			}); err != nil {
 				return err
 			}
 
-			// If there is no infrastructure Secret, e.g. WorkloadIdentity is used instead
-			// we skip the copying as it is not supported.
-			if infrastructureSecret != nil {
-				secret := &corev1.Secret{
-					ObjectMeta: metav1.ObjectMeta{Namespace: spec.Backup.CredentialsRef.Namespace, Name: spec.Backup.CredentialsRef.Name},
-				}
+			checksum = utils.ComputeSecretChecksum(secret.Data)[:8]
+		} else if apierrors.IsNotFound(originalErr) {
```

Comment:
Shouldn't this be `err` instead of `originalErr`?
The error message states that the infrastructure secret is not found, but the checked error is for the backup secret, not the infrastructure one.

```suggestion
		} else if apierrors.IsNotFound(err) {
```

Reply:
If my above suggestion is correct. Can we remove 1 level of nasting with the following diff:

```diff
diff --git a/pkg/controller/gardenletdeployer/actuator.go b/pkg/controller/gardenletdeployer/actuator.go
index 13eb14cf3a..d4e74d0da0 100644
--- a/pkg/controller/gardenletdeployer/actuator.go
+++ b/pkg/controller/gardenletdeployer/actuator.go
@@ -473,32 +473,32 @@ func (a *Actuator) reconcileSeedSecrets(ctx context.Context, obj client.Object,
 
 		infrastructureSecret, err := a.GetInfrastructureSecret(ctx)
 		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return errors.New("backup is configured to reference a secret, but there is no infrastructure secret to copy")
+			}
+
 			return err
 		}
 
 		// If there is no infrastructure Secret, e.g. WorkloadIdentity is used instead
 		// we skip the copying as it is not supported.
-		if infrastructureSecret != nil {
-			secret := &corev1.Secret{
-				ObjectMeta: metav1.ObjectMeta{Namespace: spec.Backup.CredentialsRef.Namespace, Name: spec.Backup.CredentialsRef.Name},
-			}
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{Namespace: spec.Backup.CredentialsRef.Namespace, Name: spec.Backup.CredentialsRef.Name},
+		}
 
-			if _, err := controllerutils.CreateOrGetAndStrategicMergePatch(ctx, a.GardenClient, secret, func() error {
-				secret.OwnerReferences = []metav1.OwnerReference{
-					*metav1.NewControllerRef(obj, gvk),
-				}
-				secret.Type = corev1.SecretTypeOpaque
-				secret.Data = infrastructureSecret.Data
-				delete(secret.Labels, secretStatusLabelKey)
-				return nil
-			}); err != nil {
-				return err
+		if _, err := controllerutils.CreateOrGetAndStrategicMergePatch(ctx, a.GardenClient, secret, func() error {
+			secret.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(obj, gvk),
 			}
-
-			checksum = utils.ComputeSecretChecksum(secret.Data)[:8]
-		} else if apierrors.IsNotFound(originalErr) {
-			return errors.New("backup is configured to reference a secret, but there is no infrastructure secret to copy")
+			secret.Type = corev1.SecretTypeOpaque
+			secret.Data = infrastructureSecret.Data
+			delete(secret.Labels, secretStatusLabelKey)
+			return nil
+		}); err != nil {
+			return err
 		}
+
+		checksum = utils.ComputeSecretChecksum(secret.Data)[:8]
 	} else if !allowCopying && metav1.IsControlledBy(backupSecret, obj) {
 		// backup secret was copied at an earlier stage
 		// remove the ownerReference as the controller is no longer responsible for it

```

Reply:
I think that `originalErr` is correct. This case is basically for when the user has configured a reference, but this reference is to a `WorkloadIdentity` that does not exist. Then we cannot do much and we return error.

The message should probably change to something like

```
backup is configured to reference a credential, but there is no infrastructure credential to copy
```

Reply:
OK, makes sense.